rest.go

```diff
@@ -90,39 +51,39 @@ func BuildResponse(res *http.Response) (*Response, error) {
 }
 
 // API is the main interface to the API.
-func API(request Request) (*Response, error) {
+func API(request *http.Request) (*Response, error) {
 	return DefaultClient.API(request)
 }
 
 // The following functions enable the ability to define a
 // custom HTTP Client
 
 // MakeRequest makes the API call.
-func (c *Client) MakeRequest(req *http.Request) (*http.Response, error) {
+func (c *Client) makeRequest(req *http.Request) (*http.Response, error) {
 	return c.HTTPClient.Do(req)
 }
 
 // API is the main interface to the API.
-func (c *Client) API(request Request) (*Response, error) {
-	// Add any query parameters to the URL.
-	if len(request.QueryParams) != 0 {
-		request.BaseURL = AddQueryParameters(request.BaseURL, request.QueryParams)
+func (c *Client) API(request *http.Request) (*Response, error) {
+	if request == nil {
+		return nil, errors.New("invalid request: request cannot be nil")
 	}
+	_, supportedMethod := supportedMethods[request.Method]
 
-	// Build the HTTP request object.
-	req, err := BuildRequestObject(request)
-	if err != nil {
-		return nil, err
+	if !supportedMethod {
+		return nil, errors.New("unsupported request method:" + request.Method)
+	}
+
+	if request.Header.Get("Content-Type") == "" && request.Body != nil {
+		request.Header.Set("Content-Type", "application/json")
 	}
```

Comment:
Recommendation: make `rest.Request.Body` be of `interface{}` type -- users of the package can just assign something JSON-marshallable into that field and your package will handle it. You can provide a `RawBody` field (`[]byte` is cleaner than `string` for that use case) for all other cases. If RawBody is non-nil, use that, and make no assumptions about Content-Type. Otherwise, set the Content-Type header as you do now and marshall the Body field (even if it happens to be nil -- it'll turn into a JSON `null`).